feat: 카테고리별 피드 조회 기능 구현

s9hn opened the pull request:
## 📌𝘐𝘴𝘴𝘶𝘦𝘴
- closed #33 

## 📎𝘞𝘰𝘳𝘬 𝘋𝘦𝘴𝘤𝘳𝘪𝘱𝘵𝘪𝘰𝘯
- 초기 카테고리값 호출 유즈케이스 제거
- 카테고리 전환 시 피드 업데이트 기능 구현
  - 카테고리 Ui selected 상태관리를 위해 카테고리 모델 분리 및 리팩터링
  - 카테고리 selected 지정 함수 추가
  - 카테고리 서버연결 로직 수정

## 📷𝘚𝘤𝘳𝘦𝘦𝘯𝘴𝘩𝘰𝘵
[Screen_recording_20240711_115640.webm](https://github.com/Team-WSS/WSS-Android/assets/81347125/86d749b7-2010-4f9e-9e3f-7658e493c379)


## 💬𝘛𝘰 𝘙𝘦𝘷𝘪𝘦𝘸𝘦𝘳𝘴
- 카테고리를 피드UiState와 분리했습니다.
  - 카테고리 클릭할 때마다 UiState가 업데이트되고, 필요없는 다른 로직까지 시작되는 (마치.. Recomposition..)게 마음에 안들었습니다.
  - 클릭할 때마다 뷰모델의 카테고리 상태값만 업데이트하고, 이를 따로 옵저빙하진 않습니다.(클릭 시 바로 ui에 반영하도록 구현)

s9hn commented on app/src/main/java/com/teamwss/websoso/ui/common/view/LoadingProgressIndicator.kt:
```diff
@@ -0,0 +1,35 @@
+
+    override fun onStart() {
+        super.onStart()
+        context?.let {
```
오 여기선 그냥 context가 불려와지는군요

librarywon replied:
넵 Fragment의 getContext()를 통해서 바로 받아오는게 가능하더라구요!

s9hn commented on app/src/main/java/com/teamwss/websoso/ui/main/library/LibraryViewModel.kt:
```diff
@@ -41,32 +36,40 @@ class LibraryViewModel(
     }
 
-    fun getNovels(readState: ReadState) {
+    fun setUiSateResumed() {
```
뷰모델 내의 프로퍼티를 업데이트하는 의미로 updateUiState는 어떠신가요? 뒤에 리쥼은 뭐 붙여도될거같고

librarywon replied:
set보다는 update가 더 정확한 함수명을 의미하는 것 같네요 감사합니다!

librarywon replied:
추가적으로 uistate에 onResumed를 둔 이유는

library view가 최초로 생성되었을 때와 내부 viewPager 아이템이 변경되었을 때 상태를 구분하기 위함입니다.

현재 library view에서는 tab layout을 터치하거나 스와이프 이벤트가 발생하면 새로운 통신을 진행해서 읽기상태에 따라 알맞는 ui를 보여주어야 하는 상황입니다. 

이것은 fragment view가 최소 생성되는 `Uninitialized`와는 다른 State이기 때문에 

이 상황을 정의하기 위한 State가 필요하다고 생각이 들었고 `Resumed`를 추가하여 해당 상황을 정의하고 동작을 구현하였습니다.

`observeLibraryUiState()`함수만 보면 똑같은 일을 하는데 왜 구분을 두는지 의문이 들 수 있지만 viewModel에 로직을 보면 조금 이해에 도움이 될 수 있을 것 같습니다.

다른 좋은 방법도 분명 있을 것 같은데 일단 제가 생각해본 해결방법이였습니다!

librarywon replied:
스와이프될 때 상태를 State를 Resume으로 변경합니다.

librarywon replied:
완전히 착각했었네요 onResume이라고 말했을 때 알아챘어야하는데 ㅋㅋㅋ

s9hn commented on app/src/main/java/com/teamwss/websoso/ui/main/library/LibraryViewModel.kt:
```diff
@@ -41,32 +36,40 @@ class LibraryViewModel(
+    fun getNovels() {
         viewModelScope.launch {
             runCatching {
+                _libraryUiState.value = LibraryUiState.Loading
```
스코프 밖에서 해줘도 될거같슴다

s9hn commented on app/src/main/java/com/teamwss/websoso/ui/main/library/adapter/LibraryViewPagerAdapter.kt:
```diff
@@ -7,17 +7,17 @@ import com.teamwss.websoso.data.model.LibraryUserNovelEntity
     RecyclerView.Adapter<LibraryViewPagerViewHolder>() {
 
-    private var userNovels = emptyList<LibraryUserNovelEntity>()
+    private var userNovels :List<LibraryUserNovelEntity> = libraryUserNovelsInitData
```
ㅋㅋ아 흡족하네요

librarywon replied:
리턴값 명시 필수!

s9hn commented on app/src/main/java/com/teamwss/websoso/ui/main/library/adapter/LibraryViewPagerAdapter.kt:
```diff
@@ -7,17 +7,17 @@ import com.teamwss.websoso.data.model.LibraryUserNovelEntity
 
 class LibraryViewPagerAdapter(
-    private val onItemClick: (Long) -> Unit,
+    private val userNovelClick: (userNovelId: Long) -> Unit,
```
여기도 구체적으로 명시해준거 흡족하네요

librarywon replied:
다른 레포 구경하다가 갑자기 떠올라서 구체적으로 바꿔봤습니다 ㅎㅎ

junseo511 replied:
저도 명시하는 습관 들여봐야겠어용

s9hn commented on app/src/main/java/com/teamwss/websoso/ui/main/library/adapter/LibraryViewPagerAdapter.kt:
```diff
@@ -30,4 +30,44 @@ class LibraryViewPagerAdapter(
     override fun getItemCount(): Int = 5
+
+    companion object{
+        private val libraryUserNovelsInitData = listOf(
```
추후 서버데이터로 변경될 친구인거죠?

librarywon replied:
네 맞습니다 스캘레톤 로딩과 비슷한 효과를 낸다고 볼 수 있습니다

s9hn commented on app/src/main/java/com/teamwss/websoso/util/fragment/FragmentExt.kt:
```diff
@@ -0,0 +1,19 @@
+    }
+}
+
+fun Fragment.hideLoading() {
```
여기 확장함수쪽은 나중에 팀원들한테 설명 공유바랍니다 !

librarywon replied:
넵 알겠습니다!

junseo511 replied:
저도 공유...!

s9hn reviewed (approved):
고생하셨슴다
어떤생각으로 uistate에 onResumed를 두었는지 궁금함다

m6z1 reviewed (approved):
UiState에 `onResume`와 `UnInitialized`를 따로 두신 이유에 대해서 읽어봤는데 설득당했습니다. 코드 리뷰를 따로 달지는 않았지만 코드 하나하나 다 읽어보고 이해해보려 노력했습니다.
LGTM ~ ㅎ

junseo511 commented on app/src/main/java/com/teamwss/websoso/ui/main/library/LibraryUiState.kt:
```diff
@@ -0,0 +1,9 @@
+package com.teamwss.websoso.ui.main.library
+
+sealed class LibraryUiState {
+    object Uninitialized : LibraryUiState()
+    object Loading : LibraryUiState()
+    object Resumed : LibraryUiState()
+    object Success : LibraryUiState()
+    object Error : LibraryUiState()
+}
```
분리 꼼꼼하게 해두셨군요 👍

junseo511 reviewed (approved):
서재에 누구보다 진심인 남자... 한동안 코드 안보고 있었다가 오랜만에 보네용 슬슬 다시 공부좀 잡아야겠어요 👏

s9hn committed:
refactor: 카테고리 상태 관리 로직 추가

s9hn committed:
refactor: 카테고리 상태 관리 로직 추가

s9hn committed:
refactor: 카테고리 수정

- 카테고리 모델 분리
- 카테고리 상태 업데이트 로직 수정

s9hn committed:
refactor: 카테고리 초기 selected UI 업데이트 로직 수정

- categoryUiModel 분리 및 초기세팅 시 UI 바로 적용

s9hn committed:
refactor: Websoso칩 selected UI 업데이트 로직 추가

- eventOnSingleSelectionMode 리팩터링
- 칩그룹에 초기 selected 세팅 시 singleSelectionMode 대응 로직 추가

s9hn committed:
refactor: 카테고리 서버통신 로직 리팩터링

s9hn committed:
Merge branch 'develop' of https://github.com/Team-WSS/WSS-Android into feat/33

librarywon commented on app/src/main/java/com/teamwss/websoso/data/FakeApi.kt:
```diff
@@ -40,19 +40,30 @@ object FakeApi {
         delay(500)
+
         val from = feedsRequestDto.lastFeedId.toInt()
         val to = when (feedsRequestDto.lastFeedId + 10 > Fixture.feedsResponseDto.size) {
             true -> Fixture.feedsResponseDto.size
             false -> (feedsRequestDto.lastFeedId + feedsRequestDto.size).toInt()
         }
```
a: when을 마치 if else처럼 사용하네요 정말 신선합니다

s9hn replied:
true-false는 딱 2개로 나뉘어져서 가독성 챙기기 좋은듯함다

librarywon commented on app/src/main/java/com/teamwss/websoso/data/FakeApi.kt:
```diff
@@ -40,19 +40,30 @@ object FakeApi {
+        return when {
+            to >= Fixture.feedsResponseDto.size -> Fixture.copy(
                 isLoadable = false,
                 feedsResponseDto = Fixture.feedsResponseDto.subList(from, to),
             )
-        } else Fixture.copy(
-            feedsResponseDto = Fixture.feedsResponseDto.subList(from, to),
-        )
+
+            category != null -> {
+                FeedsResponseDto(
+                    category = category,
+                    isLoadable = false,
+                    feedsResponseDto = emptyList(),
+                )
+            }
+
+            else -> Fixture.copy(
+                feedsResponseDto = Fixture.feedsResponseDto.subList(from, to),
+            )
+        }
```
a: 여기서는 더 복잡한 조건을 when으로 따지는군요 배우고 갑니다.

librarywon commented on app/src/main/java/com/teamwss/websoso/ui/feed/FeedFragment.kt:
```diff
@@ -124,11 +129,11 @@ class FeedFragment : BindingFragment<FragmentFeedBinding>(R.layout.fragment_feed
 
     private fun observeUiState() {
-        feedViewModel.uiState.observe(viewLifecycleOwner) { uiState ->
+        feedViewModel.feedUiState.observe(viewLifecycleOwner) { uiState ->
```
c: 카테고리가 제거되서 `feedUiState`로 변수 이름을 변경하셨군요 너무 좋은 것 같습니다 람다 안에 변수 이름도 바꿔주시는 것은 어떠신지

librarywon commented on app/src/main/java/com/teamwss/websoso/ui/feed/FeedViewModel.kt:
```diff
@@ -17,28 +17,54 @@ import javax.inject.Inject
+
+    init {
+        val categories: List<CategoryModel> = when (fakeUserRepository.gender) {
+            "MALE" -> "전체,판타지,현판,무협,드라마,미스터리,라노벨,로맨스,로판,BL,기타"
+            "FEMALE" -> "전체,로맨스,로판,BL,판타지,현판,무협,드라마,미스터리,라노벨,기타"
```
c: 이 성별 분류 장르는 로컬에서 가지고 있나요?

s9hn replied:
로컬에서 가진다는게 무슨의미인가요??

librarywon replied:
제가 헷갈리게 질문을 드렸네요 남자와 여자에 따른 장르 순서를 로컬에서 가지고 있는 것인가요?

s9hn replied:
지금 여기있는 스트링이 전부입니다!

librarywon commented on app/src/main/java/com/teamwss/websoso/ui/mapper/FeedMapper.kt:
```diff
@@ -2,17 +2,17 @@ package com.teamwss.websoso.ui.mapper
-        CategoryUiState(
-            category = it,
-            isSelected = it == Category.ALL,
+fun String.toPresentation(): List<CategoryModel> = split(",")
```
c: 확장함수의 이름이 너무 범용적으로 쓰이는 느낌이 있습니다. FeedMapper에서 더 구체적인 이름을 붙이는 것은 별로인가요? 물론 그렇게 하면 컨벤션을 어기는 것이 되긴합니다.

s9hn replied:
뷰모델로 이관시킵니다.

librarywon reviewed (approved):
고생하셨습니다 리뷰 확인 부탁드려여

m6z1 commented on app/src/main/java/com/teamwss/websoso/ui/common/customView/WebsosoChip.kt:
```diff
@@ -106,6 +105,15 @@ class WebsosoChip(
+   Params:
+   isSelected - The state of this chip's selected.
+    */
+    fun setWebsosoChipSelected(isSelected: Boolean) {
+        this.isSelected = isSelected
+    }
```
a: 👍 👍 👍 👍 👍 👍  따봉 50개 드립니다

librarywon replied:
개추

m6z1 commented on app/src/main/java/com/teamwss/websoso/ui/feed/FeedViewModel.kt:
```diff
@@ -17,28 +17,54 @@ import javax.inject.Inject
+
             runCatching {
-                getFeedsUseCase("전체")
+                getFeedsUseCase(selectedCategory.titleEn)
```
피드 전체 조회에서는 카테고리를 영어로 넘기네여!!? 저는 장르를 한국어로 넘기길래 서버에 여쭤봤습니다.... 코드 보니 여쭤봐야겠단 생각이 들엇네요... 감사.. ..

m6z1 reviewed (approved):
kkalkkum 하네요 👍

s9hn committed:
refactor: 카테고리 모델 메퍼 리팩터링

s9hn committed:
refactor: uiState 네이밍 수정
